### PR TITLE
Switch to SCSS

### PR DESCRIPTION
**Pull request by skjnldsv**
This is related to #1786 
Please only use this pr to discuss issues or comments about its commits.

For arguing and/or ask question, go to the issue.

### CODE REVIEW

File: lib/private/Template/SCSSCacher.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016, nextCloud, Inc.
```
Comment (MorrisJobke) [inline lib/private/Template/SCSSCacher.php]: Nope, the copyright belongs to you ;)

Comment (skjnldsv) [inline lib/private/Template/SCSSCacher.php]: Well, if you insist! :3

File: lib/private/Template/SCSSCacher.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		$subst = 'url(\'../'.$this->rootCssLoc.'/$1\')';
+		return preg_replace($re, $subst, $css);
+	}
+
+	public function getCachedSCSS() {
+		return $this->fileCache;
+	}
+}
+
+?>
```
Comment (MorrisJobke) [inline lib/private/Template/SCSSCacher.php]: No need to close this here - just remove this. This also reduces possible issues with whitespace after the closing tag that cause a different output to the browser.

**Comment by rullzer** [comment]
Please use the AppData to store stuff. Don't go writing into folders.

**Comment by LukasReschke** [comment]
> Please use the AppData to store stuff. Don't go writing into folders.

So store in AppData and use PHP to serve and set proper caching headers? We can't assume that assets is writable at the moment :)

**Comment by rullzer** [comment]
@LukasReschke exactly

**Comment by skjnldsv** [comment]
Sure! No problem! 😃

File: lib/private/Template/SCSSCacher.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016, John Molakvoæ.
+ *
+ * @author John Molakvoæ <skjnldsv@protonmail.com>
+ *
+ * @license AGPL-3.0
```
Comment (skjnldsv) [inline lib/private/Template/SCSSCacher.php]: Of course! 😃

File: core/css/share.scss [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+#defaultExpireMessage {
+	/* show message on new line */
+	display: block;
+	padding-left: 4px;
+	/* TODO: style the dropdown in a proper way - border-box, etc. */
+	width: 90%;
+}
+
+.ui-autocomplete {
```
Comment (skjnldsv) [inline core/css/share.scss]: @jancborchardt why don't we have a global definition for this? Shouldn't the dropdown use select2? :)

Comment (jancborchardt) [inline core/css/share.scss]: Opened a separate issue about that at https://github.com/nextcloud/server/issues/1831 ;)

**Comment by jancborchardt** [comment]
Btw @skjnldsv while we rewrite the whole file to new syntax and all the author info is lost, we should probably add the copyright headers at this point to each file. :)

**Comment by skjnldsv** [comment]
> we should probably add the copyright headers at this point to each file. :)
NO COPIRIT FOR U!

Yeah, I saw that. I will readd all the copyright after with a unified header so it looks good 😉

File: core/css/share.scss (hunk 4 above)
Comment (skjnldsv) [inline core/css/share.scss]: 🎩

File: lib/private/Template/SCSSCacher.php [hunk 5]
```diff
@@ -4,17 +4,18 @@
...
- * as published by the Free Software Foundation.
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
```
Comment (nickvergessen) [inline lib/private/Template/SCSSCacher.php]: I copied the header from https://github.com/nextcloud/server/blob/master/contribute/HowToApplyALicense.md

**Comment by rullzer** [comment]
@skjnldsv what is the progress here?

Any chance of breaking it up into smaller PR's As reviewing 1800 lines is kind of a pain.

**Comment by skjnldsv** [comment]
No chance at all! X)
I know this is a lot of work, but imagine what I had to do 😆 

The best way to check this pr i think will be to check the validity of the generated css, and then explore your nextcloud installation looking for mistakes (which I already did and try to check everytime I make a small change) 😕

**Comment by skjnldsv** [comment]
No chance at all! 😂 😂
I know this is a lot of work, but imagine what I had to do 😆

The best way to check this pr i think will be to check the validity of the generated css, and then explore your nextcloud installation looking for mistakes (which I already did and try to check everytime I make a small change) 😕

Concerning the progress, It's going fast I think. I'm almost done converting, and the next big thing will be the appdata integration (see original issue)

File: core/css/inputs.scss [hunk 6]
```diff
@@ -303,6 +282,12 @@ input[type="radio"].radio--white:checked:disabled + label:before {
 		margin-right: 8px;
 		vertical-align: middle;
 	}
+	.select2-search input {
```
Comment (skjnldsv) [inline core/css/inputs.scss]: capture d ecran_2016-11-08_10-34-54

**Comment by juliushaertl** [comment]
@rullzer @skjnldsv https://codeascraft.com/2015/02/02/transitioning-to-scss-at-scale/#scss-diff might be interesting to check if the resulting css code has fundamental changes.

**Comment by skjnldsv** [comment]
@juliushaertl Indeed! I wonder if I shouldn't listen to @rullzer and create separate commits.
One with the scss implementation and a basic conversion, and multiples seconds with each one a simplification. I should have done that since the beginning! 😢

**Comment by juliushaertl** [comment]
@skjnldsv Creating separate PRs should not be that hard, just create a new branch and cherry pick the commits you need from this one.

**Comment by MorrisJobke** [comment]
@skjnldsv Is this still needed?

**Comment by skjnldsv** [comment]
@MorrisJobke probably not!

**Pull request closed without merging**